src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/graphs/AdaptiveChainPruner.java

```diff
@@ -29,41 +30,73 @@ public AdaptiveChainPruner(final double initialErrorProbability, final double lo
+
+        // start from set of really obvious good chains, defined as those with an edge with multiplicity above some fraction of the
+        // maximum multiplicity edge.  Then grow the set of good chains by looking at the log odds of junctions between
+        // good chains and unknown chains.
+        final int maxEdgeWeight = graph.edgeSet().stream().mapToInt(BaseEdge::getMultiplicity).max().orElse(0);
+        final int thresholdEdgeWeight = maxEdgeWeight / 10;
+        final Set<Path<V, E>> definiteGoodChains = chains.stream()
+                .filter(chain -> chain.getEdges().stream().mapToInt(BaseEdge::getMultiplicity).max().orElse(0) > thresholdEdgeWeight)
+                .collect(Collectors.toSet());
+
+        final Set<V> terminalVerticesOfGoodChains = definiteGoodChains.stream()
+                .flatMap(c -> Stream.of(c.getFirstVertex(), c.getLastVertex())).collect(Collectors.toSet());
+
+        // initialize result to all chains that aren't definitely good and iteratively reduce this set as chains in
+        // question are found to have good log odds at junctions with good chains
+        final Set<Path<V,E>> errorChains = chains.stream().filter(c -> !definiteGoodChains.contains(c)).collect(Collectors.toSet());
+
+        // we only need to calculate log odds for questionable chains
+        final Map<Path<V,E>, Pair<Double, Double>> chainLogOdds = errorChains.stream()
                 .collect(Collectors.toMap(c -> c, c-> chainLogOdds(c, graph, errorRate)));
 
-        final Set<Path<V,E>> result = new HashSet<>(chains.size());
+        while (true) {
+            final Set<Path<V, E>> moreGoodChains = new HashSet<>();
+
+            for (final Path<V, E> chain : errorChains) {
+                if (terminalVerticesOfGoodChains.contains(chain.getFirstVertex()) && chainLogOdds.get(chain).getLeft() > logOddsThreshold) {
+                    moreGoodChains.add(chain);
+                    terminalVerticesOfGoodChains.add(chain.getLastVertex());
+                } else if (terminalVerticesOfGoodChains.contains(chain.getLastVertex()) && chainLogOdds.get(chain).getRight() > logOddsThreshold) {
+                    moreGoodChains.add(chain);
+                    terminalVerticesOfGoodChains.add(chain.getFirstVertex());
+                }
+            }
 
-        chainLogOdds.forEach((chain, lod) -> {
-            if (lod < logOddsThreshold) {
-                result.add(chain);
+            if (moreGoodChains.isEmpty()) {
+                break;
+            } else {
+                errorChains.removeAll(moreGoodChains);
             }
-        });
+        }
 
-        chains.stream().filter(c -> isChainPossibleVariant(c, graph))
-                .sorted(Comparator.comparingDouble((ToDoubleFunction<Path<V, E>>) chainLogOdds::get)
-                        .reversed().thenComparingInt(Path::length))
+        // add non-error chains to error chains if maximum number of variants has been exceeded
+        chains.stream()
```

Comment:
I'm worried about this sort, it seems to me that the LogOdds would have discrete values (since they are based on comparing counts that are descrete) so it seems likely there would be a tie here. Without a tiebreaker this seems prone to non-deterministic behavior. I would recommend either putting a relatively bulletproof set of tiebreakers in here or making all of the underlying sets Linked.

Reply:
done using your vertex lexicographical ordering suggestion as below